### PR TITLE
Kat Edges 

### PR DESCRIPTION
**Pull request by kat-perreira**
# API Muncher
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
:------------- | :-------------
How did you go about exploring the Edamam API, and how did you try querying the API? | I read the docs and used postman a lot. 
What is an API Wrapper? Why is it in `lib`? How would your project change if you needed to interact with more than one API (aka more than just the Edamam API)? | The api wrapper is a class in the lib folder. It's primary responsibility is to connect with the API. If we had more than one API we wanted to use, i'm going to guess that we would need more wrappers.
Describe your API wrapper, the methods you created in it, and why you decided to create those methods/how they were helpful | In my wrapper, I created a class method that searched for a recipe and another that showed details. I also had a bunch of helper methods down below in the private to set up the uri as an id and to create a recipe object. They were helpful because JSON is a pain to try to read and by creating a recipe object of the data received by the API, I was able to return more workable data back to the controller.
What was an edge case or failure case test you wrote for your API Wrapper? What was a nominal case? | The recipe couldnt be taken in without a uri or a label, and it couldnt be an empty string. The nominal case would be creating a Recipe object
How does VCR aid in testing an API? | It records a cassette of a response. This allows for us to test as many times as we want without making expensive calls to the API.
What is the Heroku URL of your deployed application? | https://yummymunchychowchow.herokuapp.com/   .  

* I'm sorry about the bad CSS. I tried to hit all the learning goals as best I could and styling this thing was not a priority for me.

### CODE REVIEW

File: lib/edamam_api_wrapper.rb [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+
+    new_url = CGI.escape(uri_from_id(id))
+
+    # this is passing in the new url instead of a quiery
+    url = BASE_URL + "?app_id=#{ID}" + "&app_key=#{KEY}" + "&r=#{new_url}"
+
+    response = HTTParty.get(url)
+    # call on the first of the array of hashes
+    return create_recipe(response[0])
+  end
```
Comment (droberts-sea) [inline lib/edamam_api_wrapper.rb:37]: What if the request doesn't return a recipe?

File: lib/edamam_api_wrapper.rb [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+  private
+
+  # using regex to get the last part of the uri
+  def self.id_from_uri(uri)
```
Comment (droberts-sea) [inline lib/edamam_api_wrapper.rb:42]: I like that you broke all these out as separate methods and made them private - good organization.

File: config/routes.rb [hunk 3]
```diff
@@ -0,0 +1,13 @@
+Rails.application.routes.draw do
+  # For details on the DSL available within this file, see http://guides.rubyonrails.org/routing.html
+
+  root 'search#new'
+
+  post '/search', to: 'search#index', as: 'search_create'
+
+  get ':q', to: 'search#index', as: 'search_index'
```
Comment (droberts-sea) [inline config/routes.rb:8]: This feels odd to me - having two different verbs going to the same action is a little confusing. Could you consolidate these?

File: config/routes.rb [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+
+
+  get  ':uri/show', to: 'search#show', as: 'search_show'
```
Comment (droberts-sea) [inline config/routes.rb:11]: Could you use `resources` for the index and show pages?

**Comment by droberts-sea** [comment]
# API Muncher
## What We're Looking For

Feature | Feedback
:------------- | :-------------
**Core Requirements** |
Git hygiene | yes
Comprehension questions | yes - no worries about the CSS (but thanks for letting us know)
**General** |
Rails fundamentals (RESTful routing, use of named paths) | see inline
Semantic HTML | some - structure could be much more clear
Errors are reported to the user | no - A search with no hits results in a blank page, and mangling the URI on the show page results in an exception rather than a polite error message. Getting a good handle on error handling is not just good for your users, it will also make your life easier as a developer.
API Wrapper to handle the API requests | yes
Controller testing | no
Lib testing | There are a couple, but lots of functionality is untested
Search Functionality | yes
List Functionality | yes
Show individual item functionality | yes
**Styling** |
List view shows 10 items at a time and/or has pagination | no
Attractive and usable UI | good start
**API Features** |
The App attributes Edamam | yes
The VCR cassettes do not contain the API key | yes
**External Resources** |
Link to deployed app on Heroku | yes
**Overall** | 


This is a good start. I am definitely disappointed not to see tests - the work around mocking the external dependency with VCR is an important learning goal for this assignment, and not something we'll have the opportunity to come back to. Make sure you understand how and why we do that. I would also like to see more of a focus on error handling, this is a small thing that makes a surprisingly big difference in your development as an engineer.


However, the main functionality of the site is there - searching and viewing details work great, and it is clear to me that the learning goals around working with an API and building non-model classes for a Rails app were met. It feels to me like most of the missing pieces are more a result of not having enough time than not understanding. Let me know if you disagree, or if there's specific places you'd like additional support, and keep up the hard work!
